feat: Adding batch deletion function for routing pages

FangSen9000 opened the pull request:
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

Please update this section with detailed description.

[批量删除.webm](https://user-images.githubusercontent.com/72308243/180745383-d0b5e450-5d33-44ac-a5a7-c49e5d9b4cb3.webm)
You can watch this video for a preview.

On the routing page, the user cannot delete in the state of select all or batch delete. I added related operations.
![image](https://user-images.githubusercontent.com/72308243/180339042-fb5c452b-eb9a-47cc-b03f-9af2cf19bd3a.png)


When selecting multiple or single items, a footer bar will pop up for batch deletion of table data. The deselect button will appear in the title of the table.

It also has full and anti - full pages
![image](https://user-images.githubusercontent.com/72308243/180382569-a49cfb7f-f7ad-447d-8cda-c602f16e4a91.png)

**Related issues**

#2565
#2562

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [x] Have you added corresponding test cases?
- [x] Have you modified the corresponding document?
- [x] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

FangSen9000 committed:
Update deploy-with-docker.md

FangSen9000 committed:
Update docs/en/latest/deploy-with-docker.md

Co-authored-by: Young <iskyex@outlook.com>

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update zh-CN.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

SkyeYoung commented:
@FangSen9000 Any update here?

FangSen9000 commented:
> @FangSen9000 Any update here?
I've had a day off, and I'll try to update it tonight

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update zh-CN.ts

FangSen9000 committed:
Update zh-CN.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Update service.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update zh-CN.ts

SkyeYoung commented on web/src/pages/Route/List.tsx:
```diff
@@ -118,6 +111,25 @@ const Page: React.FC = () => {
       );
     });
   };
+  
```
```suggestion
```

Please remove the extra spaces

SkyeYoung commented on web/src/pages/Route/List.tsx:
```diff
@@ -118,6 +111,25 @@ const Page: React.FC = () => {
+  const handleRemove = async (selectedRows) => {
+    const hide = message.loading('正在删除');
+    if (!selectedRows) return true;
+  
```
```suggestion
```

ditto

SkyeYoung commented:
![image](https://user-images.githubusercontent.com/48400568/179343410-a2360769-9b27-4db0-aa17-5b4e4aa64321.png)
e2e test also needs to be fixed 🤔

FangSen9000 commented:
> ![image](https://user-images.githubusercontent.com/48400568/179343410-a2360769-9b27-4db0-aa17-5b4e4aa64321.png) e2e test also needs to be fixed 🤔
I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives

FangSen9000 committed:
Update web/src/pages/Route/List.tsx

Co-authored-by: Young <iskyex@outlook.com>

FangSen9000 committed:
Update web/src/pages/Route/List.tsx

Co-authored-by: Young <iskyex@outlook.com>

SkyeYoung commented:
> > ![image](https://user-images.githubusercontent.com/48400568/179343410-a2360769-9b27-4db0-aa17-5b4e4aa64321.png) e2e test also needs to be fixed 🤔
> > I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives

Are you developing with Windows? Why not try Linux, you can use a tool such as VMWare to install. Otherwise some problems may be difficult to reproduce. 🤔

FangSen9000 commented:
> > > ![image](https://user-images.githubusercontent.com/48400568/179343410-a2360769-9b27-4db0-aa17-5b4e4aa64321.png) e2e test also needs to be fixed 🤔
> > > I really want to see if it works. I just built it in Git Bash and prompted the error code 500. This is the problem of the server. Simply using yarn start is not enough. I'm looking for node JS alternatives
> 
> Are you developing with Windows? Why not try Linux, you can use a tool such as VMWare to install. Otherwise some problems may be difficult to reproduce. 🤔

I used to use the docker recommended by the document, but I couldn't get the latest screenshot due to many small obstacles (maybe the first version can work). I wanted to use node.js yesterday starts with yarn start, which says there is a problem with the server. I really should try Linux that more people choose. I have virtual machine software, but it's a little cumbersome, so it's useless.

SkyeYoung commented:
> I used to use the docker recommended by the document, but I couldn't get the latest screenshot due to many small obstacles (maybe the first version can work). I wanted to use node.js yesterday starts with yarn start, which says there is a problem with the server. I really should try Linux that more people choose. I have virtual machine software, but it's a little cumbersome, so it's useless.

It is better to try it, it can solve a lot of possible problems

SkyeYoung commented:
![image](https://user-images.githubusercontent.com/72308243/179385289-3bfaf23f-7075-46c1-8d31-9ce29305c943.png) 

This image doesn't look like a screenshot of APISIX Dashboard, please replace it.

SkyeYoung commented:
Any update here?

FangSen9000 committed:
Update service.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update service.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Update List.tsx

SkyeYoung commented:
Did you add the English translation?

Baoyuantop commented on web/src/pages/Route/List.tsx:
```diff
@@ -542,6 +545,27 @@ const Page: React.FC = () => {
+        tableAlertOptionRender={() => {
+          return (
+            <Space size={16}>
+              <Button
```
A small suggestion, for the delete type of button we tend to use red to keep the style consistent, for this delete all the dangerous operations, it is best to add a pop-up window to confirm the second

FangSen9000 replied:
![image](https://user-images.githubusercontent.com/72308243/181669888-06518971-cea0-4075-938a-f7369207b54d.png)
In fact, in Antd UI, deletion generally prefers White，when I have spare power, I will try to add a secondary message, and then open a small patch PR

Baoyuantop replied:
Please create an issue to track this issue

Baoyuantop reviewed (commented):
Also need to add i18n files for other languages

FangSen9000 committed:
Update en-US.ts

FangSen9000 committed:
Update tr-TR.ts

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Merge branch 'apache:master' into master

Baoyuantop commented:
Please create an E2E test to cover this behavior

FangSen9000 commented:
> Please create an E2E test to cover this behavior

OK, I have opened the relevant issues and will complete them later

Baoyuantop commented:
Generally, we want tests and features in one PR

Commit:
test: add e2e test case for route batch deletion

Commit:
route batch delete test finished

Commit:
add eol

FangSen9000 committed:
Merge pull request #3 from zfkube0x1/test_for_route_batch_delete

test: add e2e test case for route batch deletion, There are some small mistakes, but it seems to have nothing to do with you. It may be that I clone the workflow of the warehouse, so I'll try to merge it.

FangSen9000 committed:
Revert "test: add e2e test case for route batch deletion"

FangSen9000 committed:
Merge pull request #4 from FangSen9000/revert-3-test_for_route_batch_delete

Revert "test: add e2e test case for route batch deletion"

SkyeYoung commented:
@FangSen9000 Is this ready to be reviewed?

FangSen9000 commented:
> @FangSen9000 Is this ready to be reviewed?

Always be ready

FangSen9000 commented:
> Generally, we want tests and features in one PR

The current new features will not cause errors in the test. I found that several powerful people have written about the test coverage, but their lint problem can't be submitted for the time being. I suggest that let those professionals open PR to study it.

SkyeYoung commented:
@FangSen9000 Hi, here's [Markdown Guide](https://www.markdownguide.org/getting-started/). Maybe u need this. 😉

SkyeYoung reviewed (changes requested):
Pls add front end e2e tests

FangSen9000 committed:
Merge branch 'apache:master' into master

FangSen9000 committed:
Create batch-delete-route.spec.js

FangSen9000 committed:
Update List.tsx

FangSen9000 commented:
> Pls add front end e2e tests

@Baoyuantop 
![image](https://user-images.githubusercontent.com/72308243/182635736-efaed1e9-5d6d-4cff-af77-61953d0c2a94.png)

I added the front-end E2E test case and deleted the successful promotion, and the detection is normal

Baoyuantop commented on web/cypress/integration/route/batch-delete-route.spec.js:
```diff
@@ -0,0 +1,124 @@
+    urisx: '/getx',
+    submitSuccess: 'Submit Successfully',
+    deleteRouteSuccess: 'Delete Route Successfully',
+    test: 'test',
+    test0: 'test0',
+    test1: 'test1',
+    test2: 'test2',
+    testx: 'testx',
+    desc0: 'desc0',
+    desc1: 'desc1',
+    desc2: 'desc2',
+    value0: 'value0',
```
Why do we need these? I didn't find out what it does.

Baoyuantop commented on web/cypress/integration/route/batch-delete-route.spec.js:
```diff
@@ -0,0 +1,124 @@
+
+  it('should delete the route', function () {
+    cy.visit('/routes/list');
+    cy.get('[type="checkbox"]').first().check();
```
This selector is not very reliable. Is there a better option? You can refer to https://docs.cypress.io/api/commands/get

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 commented:
@Baoyuantop This time, I used the selector to select the route name created, and even if more selection boxes appear in the future, it will not affect this test case. I deleted some useless selector attributes

guoqqqi commented on web/cypress/integration/route/batch-delete-route.spec.js:
```diff
@@ -0,0 +1,114 @@
+      cy.contains(data.submitSuccess);
+      cy.contains('Goto List').click();
+      cy.url().should('contains', 'routes/list');
+    }
```
It is best to assert that 3 corresponding routes were successfully created

FangSen9000 replied:
ok

guoqqqi commented on web/cypress/integration/route/batch-delete-route.spec.js:
```diff
@@ -0,0 +1,114 @@
+      force: true,
+      multiple: true,
+    });
+  });
```
It is better to add an assertion to prove that all data has been deleted

guoqqqi commented on web/src/pages/Route/List.tsx:
```diff
@@ -542,6 +545,33 @@ const Page: React.FC = () => {
+               onClick={async () => {
+               await remove(selectedRowKeys).then(() => {
+                handleTableActionSuccessResponse(
+                  `${formatMessage({ id: 'component.global.delete' })} ${formatMessage({
+                    id: 'menu.routes',
+                  })} ${formatMessage({ id: 'component.status.success' })}`,
```
It is better not to use this method for internationalization, Chinese and English are hardly compatible at the same time

guoqqqi replied:
![image](https://user-images.githubusercontent.com/72343596/183295386-383aebee-714c-47cd-a8ff-3daf99a54f75.png)
What I mean is that we shouldn't use template splicing, it leaves extra spaces in the Chinese state.
```suggestion
                  {formatMessage({ id: 'component.global.delete.routes.success' })},
```

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 commented:
@Baoyuantop @guoqqqi Do you have any questions or comments on this test case? If not, you can approve it, and then let the back-end testers debug their use cases

guoqqqi commented:
![image](https://user-images.githubusercontent.com/72343596/183295638-3c3f3bdb-da88-4c65-9b18-e50d1219ce95.png)
A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄

FangSen9000 commented:
ok




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Qi ***@***.***&gt;; 
发送时间: 2022年8月7日(星期天) 晚上10:29
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [apache/apisix-dashboard] feat: Adding batch deletion function for routing pages (PR #2502)







一个很好的提示是，通常这些评论的内容不应该被标记为由你解决，而是由为你审核的人😄

—
直接回复此邮件，在GitHub上查看，或取消订阅.
***@***.***与&gt;

FangSen9000 committed:
Apply suggestions from code review

Co-authored-by: Qi Guo <979918879@qq.com>

FangSen9000 commented:
> ![image](https://user-images.githubusercontent.com/72343596/183295638-3c3f3bdb-da88-4c65-9b18-e50d1219ce95.png) A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄

I made some mistakes after I adopted the code. Do you have any good suggestions. Is there any way to solve it? I'm debugging your update on the local machine

FangSen9000 committed:
Update List.tsx

guoqqqi commented:
> > ![image](https://user-images.githubusercontent.com/72343596/183295638-3c3f3bdb-da88-4c65-9b18-e50d1219ce95.png) A good tip is that usually the content of these reviews should not be marked as resolved by you, but by the person who reviewed it for you 😄
> 
> I made some mistakes after I adopted the code. Do you have any good suggestions. Is there any way to solve it? I'm debugging your update on the local machine

you need update the i18n file for `component.global.delete.routes.success`

FangSen9000 commented:
@guoqqqi  I got it on the local machine,  should pass him in as a variable. There was a slight problem with the syntax, which has now been resolved. Thanks for your suggestion.

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 committed:
Update List.tsx

FangSen9000 commented:
> you need update the i18n file for `component.global.delete.routes.success`
@guoqqqi You are right. This tag is a non-existent tag. I will add this tag tomorrow.
`component.global.delete.routes.success`
You created it...

FangSen9000 committed:
Update component.ts

FangSen9000 committed:
Update component.ts

FangSen9000 committed:
Update component.ts

FangSen9000 committed:
Update List.tsx

guoqqqi reviewed (approved):
LGTM

FangSen9000 commented:
@guoqqqi I have completed the changes you suggested. Check that everything is normal
@Baoyuantop @LiteSun  Are there any good questions and suggestions.

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 committed:
Update batch-delete-route.spec.js

Baoyuantop commented on web/cypress/integration/route/batch-delete-route.spec.js:
```diff
@@ -0,0 +1,165 @@
+    cy.contains(data.test0).should('not.exist');
+    cy.contains(data.test1).should('not.exist');
+    cy.contains(data.test2).should('not.exist');
+  });
```
Why do we need to test these search functions? We only need to judge that the data does not exist under https://github.com/apache/apisix-dashboard/pull/2502/files#diff-6918beb43c48e71616c503a8dffc6d9cc572f37a69d1e27f792b71fda2e5416aR117.

FangSen9000 committed:
Update batch-delete-route.spec.js

FangSen9000 commented:
These two assertions remove the name and content of the route, respectively. I think it has three advantages :


1. It can prevent only deleting the key in the route attribute, but not completely delete the route&nbsp;


2. It IS possible that THE TEST case on PAGE 2 will not BE FOUND IF there ARE more test cases later on.


3. The search function automatically refreshes the page, which can prevent timeout errors due to unrendered, and can make the test case more stable




---Original---
From: ***@***.***&gt;
Date: Mon, Aug 8, 2022 09:54 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/apisix-dashboard] feat: Adding batch deletion functionfor routing pages (PR #2502)




 
@Baoyuantop commented on this pull request.
 
 
In web/cypress/integration/route/batch-delete-route.spec.js:
 &gt; +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +    // partial match +    cy.get(selector.nameSearchInput).clear().type(data.test2); +    cy.contains('Search').click(); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +    // no match +    cy.get(selector.nameSearchInput).clear().type(data.testx); +    cy.contains('Search').click(); +    cy.contains(data.test0).should('not.exist'); +    cy.contains(data.test1).should('not.exist'); +    cy.contains(data.test2).should('not.exist'); +  });  
Why do we need to test these search functions? We only need to judge that the data does not exist under https://github.com/apache/apisix-dashboard/pull/2502/files#diff-6918beb43c48e71616c503a8dffc6d9cc572f37a69d1e27f792b71fda2e5416aR117.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;
